plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/machine/backup/MachineBackupManager.java

```diff
@@ -89,7 +120,22 @@ public void backupWorkspace(final String workspaceId,
     public void backupWorkspaceAndCleanup(final String workspaceId,
                                           final String srcPath,
                                           final String srcAddress) throws ServerException {
-        backupWorkspace(workspaceId, srcPath, srcAddress, true);
+        ReentrantLock lock = workspacesBackupLocks.get(workspaceId);
+        if (lock != null) {
+            lock.lock();
+            try {
+                if (workspacesBackupLocks.get(workspaceId) == null) {
+                    // it is possible to reach here if invoke this method again while previous one is in progress
+                    // should never happen
+                    LOG.error("Backup with cleanup of the workspace {} was invoked several times simultaneously", workspaceId);
+                    return;
+                }
+                backupWorkspace(workspaceId, srcPath, srcAddress, true);
+            } finally {
+                workspacesBackupLocks.remove(workspaceId);
+                lock.unlock();
+            }
+        }
```

Comment:
Please add the same as in https://github.com/codenvy/codenvy/pull/425/files#diff-6fa2b3319d5dc836fd809bbba1ed5605R106